src/workerd/jsg/wrappable.h

```diff
@@ -23,6 +23,41 @@ using kj::uint;
 
 class GcVisitor;
 class HeapTracer;
+class IsolateBase;
+
+class IsolatePtr {
+  // IsolatePtr is a safe(r) wrapper around a raw v8::Isolate pointer. Will ensure
+  // that a nullptr is returned after the Isolate is Disposed or the IsolatePtr
+  // is moved away.
+public:
+  IsolatePtr() = default;
+  IsolatePtr(decltype(nullptr)) {}
+  IsolatePtr(v8::Isolate* isolate);
+  IsolatePtr(IsolatePtr& other);
+  IsolatePtr(IsolatePtr&&) = default;
+
+  inline operator v8::Isolate*() const { return get(); }
+
+  IsolatePtr& operator=(IsolatePtr& other);
+  IsolatePtr& operator=(IsolatePtr&& other) = default;
+
+  v8::Isolate* get() const;
+  inline v8::Isolate* operator->() const { return get(); }
+  inline void clear() { impl = nullptr; }
+
+private:
+  struct Impl : public kj::Refcounted {
```

Comment:
I think this is going to need to be `AtomicRefcounted`, because `Data` is allowed to be held and destroyed outside the lock. Hence the logic in `destroy()` which detects if the lock is held and, if not, carefully places the underlying handle in a queue to be destroyed later when the lock is held. But we cannot easily use that queue to handle decrementing the refcount because the whole point is that the isolate might already be destroyed, so there's no queue to send to.

Atomic refcounting, though, is expensive. Atomics are really slow. So this guardrail may have a non-trivial cost.